pkg/impersonation/impersonation.go

```diff
@@ -323,3 +344,84 @@ func (i *Impersonator) waitForServiceAccount(sa *corev1.ServiceAccount) (*corev1
 	}
 	return ret, nil
 }
+
+func getUser(clusterContext *config.UserContext, userInfo user.Info, groupName string) (user.Info, error) {
+	u, err := clusterContext.Management.Management.Users("").Controller().Lister().Get("", userInfo.GetUID())
+	if err != nil {
+		return &user.DefaultInfo{}, err
+	}
+
+	groups := []string{"system:authenticated", "system:cattle:authenticated"}
+	if groupName != "" {
+		groups = append(groups, groupName)
+	}
+	attribs, err := clusterContext.Management.Management.UserAttributes("").Controller().Lister().Get("", userInfo.GetUID())
+	if err != nil && !apierrors.IsNotFound(err) {
+		return &user.DefaultInfo{}, err
+	}
+	if attribs != nil {
+		for _, gps := range attribs.GroupPrincipals {
+			for _, groupPrincipal := range gps.Items {
+				if !isInList(groupPrincipal.Name, groups) {
+					groups = append(groups, groupPrincipal.Name)
+				}
+			}
+		}
+	}
+	sort.Strings(groups)
+
+	extras := getUserExtras(userInfo)
+	if !isInList(u.Username, extras["username"]) {
+		extras["username"] = append(extras["username"], u.Username)
+	}
+	sort.Strings(extras["username"])
+	for _, principalid := range u.PrincipalIDs {
+		if !isInList(principalid, extras["principalid"]) {
+			extras["principalid"] = append(extras["principalid"], principalid)
+		}
+	}
+	sort.Strings(extras["principalid"])
+
+	user := &user.DefaultInfo{
+		UID:    u.GetName(),
+		Name:   u.Username,
+		Groups: groups,
+		Extra:  extras,
+	}
+	return user, nil
+}
+
+func removeEmpty(list []string) []string {
+	ret := make([]string, 0, cap(list))
+	for _, s := range list {
+		if s != "" {
+			ret = append(ret, s)
+		}
+	}
+	return ret
+}
+
+func getUserExtras(userInfo user.Info) map[string][]string {
+	extras := userInfo.GetExtra()
+	if extras == nil {
+		return make(map[string][]string)
+	}
+	if _, ok := extras["username"]; ok {
+		extras["username"] = removeEmpty(extras["username"])
+		sort.Strings(extras["username"])
```

Comment:
I don't believe we need to sort here or below for `extras["principalid"]` as we sort again after adding extra item(s) to these lists in the calling func.

Reply:
done